service/collector/cluster.go

```diff
@@ -0,0 +1,135 @@
+	"fmt"
+
+	"github.com/giantswarm/apiextensions-application/api/v1alpha1"
+	"github.com/giantswarm/k8sclient/v7/pkg/k8sclient"
+	"github.com/giantswarm/k8smetadata/pkg/label"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/prometheus/client_golang/prometheus"
+	k8slabels "k8s.io/apimachinery/pkg/labels"
+	capiv1alpha4 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/cluster-apps-operator/pkg/project"
+)
+
+const (
+	labelClusterID = "cluster_id"
+)
+
+var (
+	danglingApps *prometheus.Desc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, subsystem, "dangling_apps"),
+		"Number of apps not yet deleted for a terminating cluster.",
+		[]string{
+			labelClusterID,
+		},
+		nil,
+	)
+)
+
+type ClusterConfig struct {
+	K8sClient k8sclient.Interface
+	Logger    micrologger.Logger
+}
+
+type Cluster struct {
+	context   context.Context
+	k8sClient k8sclient.Interface
+	logger    micrologger.Logger
+}
+
+func NewCluster(config ClusterConfig) (*Cluster, error) {
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	np := &Cluster{
+		context:   context.Background(),
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+	}
+
+	return np, nil
+}
+
+func (c *Cluster) Collect(ch chan<- prometheus.Metric) error {
+	var clusterList capiv1alpha4.ClusterList
+	{
+		err := c.k8sClient.CtrlClient().List(
+			c.context,
+			&clusterList,
+		)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	for _, cl := range clusterList.Items {
+		if cl.DeletionTimestamp.IsZero() || !hasFinalizer(cl.GetFinalizers()) {
+			continue
+		}
+
+		dangling, err := c.getNumberOfApps(cl.GetName(), cl.GetNamespace())
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		ch <- prometheus.MustNewConstMetric(
+			danglingApps,
+			prometheus.GaugeValue,
+			float64(dangling),
+			cl.GetName(),
+		)
+
+	}
+
+	return nil
+}
+
+func (c *Cluster) Describe(ch chan<- *prometheus.Desc) error {
+	ch <- danglingApps
+
+	return nil
+}
+
+func (c *Cluster) getNumberOfApps(name, namespace string) (int, error) {
+	var appList v1alpha1.AppList
+	{
+		var err error
+
+		selector, err := k8slabels.Parse(fmt.Sprintf("%s=%s,%s!=%s", label.Cluster, name, label.ManagedBy, project.Name()))
+		if err != nil {
+			return -1, microerror.Mask(err)
+		}
+
+		o := client.ListOptions{
+			Namespace:     namespace,
+			LabelSelector: selector,
+		}
+
+		err = c.k8sClient.CtrlClient().List(c.context, &appList, &o)
+		if err != nil {
+			return -1, microerror.Mask(err)
+		}
+	}
+
+	return len(appList.Items), nil
+}
+
+func hasFinalizer(finalizers []string) bool {
+	for _, f := range finalizers {
+		if f == fmt.Sprintf("operatorkit.giantswarm.io/%s-cluster-controller", project.Name()) {
```

Comment:
Do we not have some `key` for the full finalizer already somewhere?

Reply:
Good point @MarcelMue , I don't know but will check that

Reply:
We don't have a key func because the finalizer is configured via the cluster controller [here](https://github.com/giantswarm/cluster-apps-operator/blob/8112f5c82dcb6bdbc6921b10f3486b1ca24e08aa/service/controller/cluster.go#L69-L71).

It makes sense to add one. 👍

Reply:
@rossf7 @MarcelMue couldn't we export this one? https://github.com/giantswarm/operatorkit/blob/04ea5c970d91bf94680a0fd5a7be299a655b1aab/pkg/controller/finalizer.go#L249

Reply:
@ljakimczuk Yes, that would do it.

I think it makes sense. We might have some key funcs in other operators that it could replace.

Reply:
Ok, have it here: https://github.com/giantswarm/operatorkit/pull/598